python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,175 @@
+    +------+----+---------------+                                                   
+    |course|year|median_earnings|
+    +------+----+---------------+
+    |dotNET|2012|           7500|
+    |  Java|2012|          20000|
+    |  Java|2013|          30000|
+    |dotNET|2013|          48000|
+    +------+----+---------------+
+    <BLANKLINE>
+    >>> def twoKeyYearlyMedian(_, vals):
+    ...     return pd.DataFrame.from_records([(int(vals["earnings"].median()),)])
+    >>> newSchema = StructType([df.schema["earnings"]])
+    >>> gapply(df.groupBy("course", "year"), twoKeyYearlyMedian, newSchema, "earnings") \
+            .orderBy("earnings").show()
+    +------+----+--------+
+    |course|year|earnings|
+    +------+----+--------+
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    """
+    import pandas as pd
+    minPandasVersion = '0.7.1'
+    if LooseVersion(pd.__version__) < LooseVersion(minPandasVersion):
+        raise ImportError('Pandas installed but version is {}, {} required'
+                          .format(pd.__version__, minPandasVersion))
+
+    # Do a null aggregation to retrieve the keys first (should be no computation)
+    # Also consistent with spark.sql.retainGroupColumns
+    keySchema = grouped_data.agg({}).schema
+    keyCols = grouped_data.agg({}).columns
+
+    if not cols:
+        # Extract the full column list with the parent df
+        javaDFName = "org$apache$spark$sql$RelationalGroupedDataset$$df"
+        parentDF = java_gateway.get_field(grouped_data._jgd, javaDFName)
+        allCols = DataFrame(parentDF, None).columns
+        keyColsSet = set(keyCols)
+        cols = [col for col in allCols if col not in keyColsSet]
+
+    if "*" in cols:
+        raise ValueError("cols expected to contain only singular columns")
+
+    if len(set(cols)) < len(cols):
+        raise ValueError("cols expected not to contain duplicate columns")
+
+    if not isinstance(schema, StructType):
+        raise ValueError("output schema should be a StructType")
+
+    inputAggDF = grouped_data.agg({col: 'collect_list' for col in cols})
+    # Recover canonical order (aggregation may change column order)
+    cannonicalOrder = chain(keyCols, (inputAggDF['collect_list(' + col + ')'] for col in cols))
+    inputAggDF = inputAggDF.select(*cannonicalOrder)
+
+    # Wraps the user-proveded function with another python function, which prepares the
+    # input in the form specified by the documentation. Then, once the function completes,
+    # this wrapper prepends the keys to the output values and converts from pandas.
+    def pandasWrappedFunc(*args):
+        nvals = len(cols)
+        keys, collectedCols = args[:-nvals], args[-nvals:]
+        paramKeys = tuple(keys)
+        if len(paramKeys) == 1: paramKeys = paramKeys[0]
+        valuesDF = pd.DataFrame.from_dict(
```

Comment:
It will help the reader if you use something like `PDF` as a suffix for pandas dataframes, to clearly separate them from the spark dataframes.